product_end_of_life/models/product.py

```diff
@@ -0,0 +1,187 @@
+        update_result = []
+        today = fields.Date.today()
+        if 'source_from' in context and \
+                context.get('source_from') == 'purchase_order':
+            if result:
+                for res in result:
+                    for product in self.browse(res[0]):
+                        # Exclude expired products
+                        if product.eol_date and product.eol_date <= today:
+                            continue
+                        else:
+                            # Design product code, name
+                            def _name_get(d):
+                                name = d.get('name', '')
+                                code = \
+                                    context.get('display_default_code', True) \
+                                    and d.get('default_code', False) or False
+                                if code:
+                                    name = '[%s] %s' % (code, name)
+                                return (d['id'], name)
+                                # prepare product attributes
+                            variant = ", ".join([v.name for v in
+                                                 product.attribute_value_ids])
+                            name = variant and "%s (%s)" % \
+                                (product.name, variant) or product.name
+                            mydict = {
+                                'id': product.id,
+                                'name': name,
+                                'default_code': product.default_code,
+                            }
+                            update_result.append(_name_get(mydict))
+        else:
+            update_result = result
+        return update_result
+
+    @api.model
+    def search_read(self, domain=None, fields=None, offset=0, limit=None,
+                    order=None):
+        res = super(ProductVariants, self).search_read(domain=domain,
+                                                       fields=fields,
+                                                       offset=offset,
+                                                       limit=limit,
+                                                       order=order)
+        today = str(datetime.today().date())
+        context = self._context or {}
+        if 'source_from' in context and \
+                context.get('source_from') == 'purchase_order':
+            if not((res and domain) is None) and len(domain) > 0:
+                params = self.env.context and \
+                        self.env.context.get('params', False) or False
+                action = params and params.get('action', False) or False
+                if not action:
+                    final_res = []
+                    for product_product in res:
+                        if 'id' in product_product:
+                            for product in self.env['product.product'].browse(
+                                    product_product.get('id')):
```

Comment:
Same problem than before, do not browse inside the loop.

Create a recordset first, then loop
```python
product_ids = [product['id'] for product in res if product.get('id')]
products = self.env['product.product'].browse(product_ids)
```